_posts/2020-09-21-centralised-schema-management-microservices.md

```diff
@@ -0,0 +1,209 @@
+
+Yes.
+
+There are things like [Confluent Schema Registry](https://docs.confluent.io/current/schema-registry/index.html) which provides tight integration with Apache Kafka but also works with anything else that can consume JSON Schema, Avro or Protobuf. I don't have any hands-on experience with it, but looks like it provides most, but not all of the functionality described above.
+
+In any case, we rolled our own because we could not find anything fitting our requirements. Besides, this post is already too long to include a market survey.
+
+# How we did it (and why)
+
+As I said above, in our project ([Velho](https://vayla.fi/hankkeet/digitalisaatiohanke/tieverkon-kunnonhallinta/velho-allianssi), for the [Finnish Transport Infrastructure Agency](https://vayla.fi/)) we ended up creating our own, custom centralized schema management solution. This was due to several reasons:
+
+- Our project has to produce a long-lived system, and our architectural solution to that was a "midi-service" (i.e. less granular than micro, but definitely not a monolith) approach. So we would have a moderately large amount of independent services, each with their own responsibility of data storage etc.
+- To allow for a [ship of Theseus](https://en.wikipedia.org/wiki/Ship_of_Theseus) -type evolution path for the entire system, our internal architecture would explicitly and consistently support a polyglot implementation. This means every service chooses its own tech stack, and disallows the use of any integration technologies specific to any single language or platform. To keep ourselves honest we [dogfood](https://en.wikipedia.org/wiki/Eating_your_own_dog_food), i.e. consume internally the same protocols and APIs we provide externally.
+- The number of data types our system needs to manage is around 100, of varying complexity, each with their own properties and quirks. They're split into multiple collections and each collection is managed by a different service.
+- The above made the need for a CSM evident early on.
+- The existing options did not feel suitable. Granted, we didn't do any really comprehensive market research. So sue me.
+
+## Our stack
+
+[AWS native](https://aws.amazon.com/architecture/well-architected/). Containerized with AWS Fargate, FaaS with AWS Lambda. Multiple independent SQL databases (Aurora PostgreSQL). Elasticsearch. Redis. S3. AWS API Gateway. Infrastructure-as-Code via CloudFormation.
+
+Languages: [Clojure](http://clojure.org) backend services. [ClojureScript](https://clojurescript.org/) frontend with [Reagent](https://reagent-project.github.io/), [Re-frame](https://github.com/day8/re-frame) and [Web Components](https://www.webcomponents.org/introduction). Lambdas in Clojure, Python and Javascript.
+
+## What we use CSM for
+
+- We validate incoming data (APIs, import jobs, etc) using CSM-provided schemas.
+  - Inbound data using old schema versions is supported, and we automatically transform it into the latest version when needed. 
```

Comment:
How can this be done automatically for example when the new schema has a new absolutely required field that the old schema doesn't have? The example you've given is a trivial field name change with no change in data, so as an example it's kind of uninteresting.